File: C/2_add_two_nums.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    int add_val = 0;
    struct ListNode* lret_head = NULL;
    struct ListNode* cur_node_lret = NULL;

    while (l1 != NULL || l2 != NULL || add_val != 0) {
        if (cur_node_lret == NULL) {
            cur_node_lret = (struct ListNode*)malloc(sizeof(struct ListNode));
            lret_head = cur_node_lret;
        } else {
            cur_node_lret->next = (struct ListNode*)malloc(sizeof(struct ListNode));
            cur_node_lret = cur_node_lret->next;            
        }
        
        memset(cur_node_lret, 0, sizeof(struct ListNode));
        
        if (l1 != NULL) {
            cur_node_lret->val += l1->val;
            l1 = l1->next;
        }

        if (l2 != NULL) {
            cur_node_lret->val += l2->val;
            l2 = l2->next;
        }

        cur_node_lret->val += add_val;
        add_val = cur_node_lret->val / 10;
        cur_node_lret->val = cur_node_lret->val % 10;        
    }

    return lret_head;
}

File: Python/2_add_two_nums.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if l1 is None:
            if l2 is None:
                return None
            else:
                return l2
        else:
            if l2 is None:
                return l1
        
        sum = l1.val + l2.val
        ca = sum / 10
        re = sum % 10
        head_node = ListNode(re)
        cnode = head_node
        
        while True:
            if l1.next and l2.next:
                l1 = l1.next
                l2 = l2.next
                sum = l1.val + l2.val + ca
                ca = sum / 10
                re = sum % 10
                nnode = ListNode(re)
                cnode.next = nnode
                cnode = nnode
            elif l1.next is None:
                if l2.next is None:
                    if ca is 0:
                        return head_node
                    else:
                        nnode = ListNode(ca)
                        cnode.next = nnode
                        return head_node
                else:
                    while l2.next:
                        l2 = l2.next
                        sum = l2.val + ca
                        ca = sum / 10
                        re = sum % 10
                        nnode = ListNode(re)
                        cnode.next = nnode
                        cnode = nnode
                    if ca:
                        nnode = ListNode(ca)
                        cnode.next = nnode
                    return head_node
            elif l2.next is None:
                if l1.next is None:
                    if ca is 0:
                        return head_node
                    else:
                        nnode = ListNode(ca)
                        cnode.next = nnode
                        return head_node
                else:
                    while l1.next:
                        l1 = l1.next
                        sum = l1.val + ca
                        ca = sum / 10
                        re = sum % 10
                        nnode = ListNode(re)
                        cnode.next = nnode
                        cnode = nnode
                    if ca:
                        nnode = ListNode(ca)
                        cnode.next = nnode
                    return head_node
        